Compatibility check and publish response code

martinlukas84 opened the pull request:
## Describe your changes
This PR adds the compatibility check according to the standard IEC61851-23 CC.5.6.
Furthermore, adding the response code to the feedback to get this information available for other modules.
This is helpful for process view in PLC, HMI etc.

## Issue ticket number and link

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

martinlukas84 committed:
The standard IEC61851-23 CC.5.6 requires a compatibility check at state ChargeParameterDiscovery.
This will be added here.

Signed-off-by: Martin Lukas <martin.lukas@chargebyte.com>

martinlukas84 committed:
To get the response code available for other modules in EVerest it's send via feedback to the Iso module.

Signed-off-by: Martin Lukas <martin.lukas@chargebyte.com>

SebaLukas commented on src/iso15118/d20/state/dc_charge_parameter_discovery.cpp:
```diff
@@ -42,6 +42,81 @@ template <> void convert(BPT_DC_ModeRes& out, const d20::DcTransferLimits& in) {
     }
 }
 
+bool handle_compatibility_check(const d20::DcTransferLimits& evse_dc_limits, const session::feedback::DcMaximumLimits ev_max_limits) {
```
ev_max_limits is a const copy -> should be a const reference

SebaLukas replied:
Also why using `session::feedback::DcMaximumLimits`? This should be `DC_ModeReq` or `BPT_DC_ModeReq` as a template function.

martinlukas84 replied:
This is used in the feed method to get the limits out of the req. And the idea wa to forward it to the compatibility_check method. But i will take look to improve this.

SebaLukas commented on src/iso15118/d20/state/dc_charge_parameter_discovery.cpp:
```diff
@@ -42,6 +42,81 @@ template <> void convert(BPT_DC_ModeRes& out, const d20::DcTransferLimits& in) {
+    // the EVerest environment (power supply, grid), here we assume that both are the same
+
+    bool compatiblity_flag = true;
+    const auto ev_max_limits_cpd = ev_max_limits;
```
Why did you add another copy here? Why not using the argument `ev_max_limits` as const ref

martinlukas84 replied:
up to now i was not sure how to handle the CPD and RATED index. I thought to work with new named vars to keep a kind sort of readability. But when we now get rid of the CPD and RATED index it can be removed or implemented in other way.

SebaLukas replied:
If you like to use variables for this, it's better to use const ref instead of copying each variable.

SebaLukas commented on src/iso15118/d20/state/dc_charge_parameter_discovery.cpp:
```diff
@@ -42,6 +42,81 @@ template <> void convert(BPT_DC_ModeRes& out, const d20::DcTransferLimits& in) {
+
+
+    // CC.5.6 2.a
+    if (ev_max_limits_cpd.voltage <= 500.f) {
```
Too many magic numbers. Add constexpr for every magic number within the function

SebaLukas commented on src/iso15118/d20/state/dc_charge_parameter_discovery.cpp:
```diff
@@ -42,6 +42,81 @@ template <> void convert(BPT_DC_ModeRes& out, const d20::DcTransferLimits& in) {
+    // CC.5.6 2.d-f
+    const float evse_voltage_min_rated = evse_voltage_min_cpd;
+    const float evse_current_min_rated = evse_current_min_cpd;
+    const float evse_power_min_rated = evse_power_min_cpd;
```
Why do you need so many copies?

martinlukas84 replied:
The same thought as above.

SebaLukas commented on include/iso15118/detail/d20/state/dc_charge_parameter_discovery.hpp:
```diff
@@ -13,4 +13,7 @@ message_20::DC_ChargeParameterDiscoveryResponse
                const d20::DcTransferLimits& dc_limits);
 
+bool handle_compatibility_check(const d20::DcTransferLimits& evse_dc_limits,
+                                const session::feedback::DcMaximumLimits ev_max_limits);
```
This function should be local function in `dc_charge_parameter_discovery.cpp`

SebaLukas commented on src/iso15118/d20/state/dc_charge_parameter_discovery.cpp:
```diff
@@ -122,11 +197,17 @@ Result DC_ChargeParameterDiscovery::feed(Event ev) {
+        auto res = handle_request(*req, m_ctx.session, m_ctx.session_config.powersupply_limits);
+
+        // Do compatibility check IEC61851-23 CC.5.6
+        if (not handle_compatibility_check(m_ctx.session_config.dc_limits, dc_max_limits)) {
```
This check function should be called in the `handle_request()` and not here. Then your changes can also be tested. And the res can stay const.
An unit test is also missing

SebaLukas commented on src/iso15118/d20/state/dc_charge_parameter_discovery.cpp:
```diff
@@ -42,6 +42,81 @@ template <> void convert(BPT_DC_ModeRes& out, const d20::DcTransferLimits& in) {
+    // In IEC 61851-23-3 a compatibility check is required
+    // There are voltage, current and power values indexed with CPD and RATED.
+    // The CPD is the communicated value in DC_ChargeParameterDiscoveryRes and RATED the value coming from
+    // the EVerest environment (power supply, grid), here we assume that both are the same
```
For readability I think you can remove cpd and rated. I think it is obvious which parameter comes from where (evse or ev)

SebaLukas replied:
After reading CC.5.6 more carefully, I now understand where the CPD and Rated names come from. Nevertheless, they make less sense to me in code.

martinlukas84 committed:
Rework after review

Signed-off-by: Martin Lukas <martin.lukas@chargebyte.com>

martinlukas84 committed:
A good case test failed where the EV max voltage is below 500V. This concerns part CC.5.6 2a in standard
IEC61851-23. The test case defines a powersupply voltage of 900V. But this must be meet
the requirement from mentioned part min(V_ev +50, V_evse, 500).
So here the V_Evse must be <= 450V.

Signed-off-by: Martin Lukas <martin.lukas@chargebyte.com>

martinlukas84 committed:
Add unit test for MCS and MCS_BPT for testing compatibility check.
Fix wrong parameter usage in compatibility check.

Signed-off-by: Martin Lukas <martin.lukas@chargebyte.com>

SebaLukas commented:
I took a closer look at `CC.5.6` in `IEC61851-23:2023` and don't seem to interpret it the same way you do.
I understand section `CC.5.6. 2) a) - c)` to mean rather that the charger should adjust the parameters for `ChargeParameterDiscoveryRes` and not check whether the values are within the value range.
For me, it would be fatal if a charger ended the session with an error every time just because the max EVSE voltage is 1000V and the EV only specifies a maximum voltage of 500V.
Therefore, `a) - c)` should adjust the parameters of the ChargeParameterDiscoveryRes accordingly.

For me the g) section also describes a check in which the charger changes the `ResponseCode`, `Processing`, and `StatusCode` accordingly.

In addition, `ChargeParameterDiscoveryRes` or any other res message should not have any values in the event of an error, if possible. I saw that in the newly added unit tests.

SebaLukas commented:
I would close this one because of the libiso migration into everest-core.
Feel free to open a PR in everest-core :)
